src/main/java/org/apache/mesos/hdfs/state/PersistentState.java

```diff
@@ -42,26 +56,171 @@ public void setFrameworkId(FrameworkID frameworkId) throws InterruptedException,
+      }
+    }
+    return deadJournalHosts;
   }
 
-  public void setNamenode(Node node) throws InterruptedException, ExecutionException, IOException {
-    set(NAMENODE_KEY, node);
+  public List<String> getDeadNameNodes() {
+    HashMap<String, String> nameNodes = getNameNodes();
+    Set<String> nameHosts = nameNodes.keySet();
+    List<String> deadNameHosts = new ArrayList<>();
+
+    for (String nameHost : nameHosts) {
+      if (nameNodes.get(nameHost) == null) {
+        deadNameHosts.add(nameHost);
+      }
+    }
+    return deadNameHosts;
   }
 
-  public Set<Node> getNodes() throws InterruptedException, ExecutionException, IOException,
-      ClassNotFoundException {
-    Set<Node> nodes = get(NODES_KEY);
-    if (nodes == null)
-      nodes = new HashSet<Node>();
-    return nodes;
+  public List<String> getDeadDataNodes() {
+    HashMap<String, String> dataNodes = getDataNodes();
+    Set<String> dataHosts = dataNodes.keySet();
+    List<String> deadDataHosts = new ArrayList<>();
+
+    for (String dataHost : dataHosts) {
+      if (dataNodes.get(dataHost) == null) {
+        deadDataHosts.add(dataHost);
+      }
+    }
+    return deadDataHosts;
+  }
+
+  // TODO (nicgrayson) add tests with in-memory state implementation for zookeeper
+  public HashMap<String, String> getJournalNodes() {
+    return getHashMap(JOURNALNODES_KEY);
   }
 
-  public void setNodes(Set<Node> nodes) throws InterruptedException, ExecutionException,
-      IOException {
-    set(NODES_KEY, nodes);
+  public HashMap<String, String> getNameNodes() {
+    return getHashMap(NAMENODES_KEY);
+  }
+
+  public HashMap<String, String> getDataNodes() {
+    return getHashMap(DATANODES_KEY);
+  }
+
+  public Collection<String> getAllTaskIds() {
+    HashMap<String, String> allTasksIds = getJournalNodes();
+    allTasksIds.putAll(getNameNodes());
+    allTasksIds.putAll(getDataNodes());
+    return allTasksIds.values();
+  }
+
+  public void addHdfsNode(Protos.TaskID taskId, String hostname, String taskName) {
+    switch (taskName) {
+      case HDFSConstants.NAME_NODE_ID :
+        HashMap<String, String> nameNodes = getNameNodes();
+        nameNodes.put(hostname, taskId.getValue());
+        System.out.println("Saving the name node " + hostname + " " + taskId.getValue());
+        setNameNodes(nameNodes);
+        break;
+      case HDFSConstants.JOURNAL_NODE_ID :
+        HashMap<String, String> journalNodes = getJournalNodes();
+        journalNodes.put(hostname, taskId.getValue());
+        setJournalNodes(journalNodes);
+        break;
+      case HDFSConstants.DATA_NODE_ID :
+        HashMap<String, String> dataNodes = getDataNodes();
+        dataNodes.put(hostname, taskId.getValue());
+        setDataNodes(dataNodes);
+        break;
+      case HDFSConstants.ZKFC_NODE_ID :
+        break;
+      default :
+        log.error("Task name unknown");
+    }
+  }
+
+  // TODO (elingg) optimize this method/ Possibly index by task id instead of hostname
```

Comment:
Well, I think you want to index by hostname for faster lookup when you get an offer to see "is there already a JN/NN/DN running on that hostname?" But if you had removeTask(slaveId, taskId), then you could index into the maps instead of iterating.